src/main/java/htsjdk/samtools/util/AsyncReadTaskRunner.java

```diff
@@ -0,0 +1,274 @@
+package htsjdk.samtools.util;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.*;
+
+/**
+ * Helper class for performing asynchronous reading.
+ * This class handles the asynchronous task scheduling.
+ * Asynchronous tasks are split into single blocking read ahead task (performReadAhead())
+ * and any number of non-blocking transform tasks (transform()).
+ * Synchronisation of the underlying stream/iterator is handled by ensuring that performReadAhead() is not executed
+ * by more than one thread concurrently.
+ * As this represents a serial bottleneck limiting palatalisation,
+ * implementations should ensure that as much computation work (such as decompression or record parsing) as possible
+ * is performed in transform().
+ */
+public abstract class AsyncReadTaskRunner<T, U> {
+    private static final Log log = Log.getInstance(AsyncReadTaskRunner.class);
+    private static Executor nonblockingThreadpool = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors(), r -> {
+        Thread t = Executors.defaultThreadFactory().newThread(r);
+        t.setName("htsjdk-asyncio-nonblocking");
+        t.setDaemon(true);
+        return t;
+    });
+
+    private static Executor blockingThreadpool = Executors.newCachedThreadPool(r -> {
+        Thread t = Executors.defaultThreadFactory().newThread(r);
+        t.setName("htsjdk-asyncio-blocking");
+        t.setDaemon(true);
+        return t;
+    });
+
+    /**
+     * Thread pool for non-blocking computational tasks that do not perform any blocking operations.
+     */
+    public static Executor getNonBlockingThreadpool() { return nonblockingThreadpool; }
+    /**
+     * Thread pool for blocking IO tasks. Tasks scheduled here should do as little computational work
+     * as possible.
+     */
+    public static Executor getBlockingThreadpool() { return blockingThreadpool; }
+    public static void setNonblockingThreadpool(Executor nonblockingThreadpool) {

```

Comment:
As it doesn't really matter much which thread pool tasks get scheduled to I went with the simple/conservative approach that existing tasks were free to complete on the old thread pool, with new ones getting scheduled on the new one. This approach also avoids having to synchronise on the thread pool object as well as avoiding having to block/throw in the set() method.

Since the operations in progress are free to complete on the existing thread pool, there's no need to throw.